### PR TITLE
Fix gas estimation for Ethereum transactions with historical block

### LINKED ISSUE #14064: Gas estimation for Ethereum transactions is below acceptable range
### Description

When comparing historical contract results with the gas estimations from the `/contract/call` endpoint it seems that the Ethereum transactions show consistently a result that is below the acceptable range (even below the actual gas used).

An example, ran against testnet is:

```
Out of tolerance for hash=0xb5a806adecaf8a6c6fa2e333f9b57b497c217a1affebb9b8ca5110126c763f6f: estimated=347576 gas_used=398389 overhead=-12.75% (expected 2-20%)
```

The request body:
```
{
    "from": "0xeb5c5a71bff6ba9bff2c887155ed7283ff79696f",
    "to": "0x000000000000000000000000000000000096785d",
    "data": "0xa872c9396c9ff7d1c5e5b81cff2e9b4e9eb67d09253ddd7ae3b6a2fadeb005db2e48bcd5000000000000000000000000000000000000000000000000000000000000012e000000000000000000000000000000000000000000000000000000006a7c1c920000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000764d90d2aed00000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000003c000000000000000000000000000000000000000000000000000000000017011f400000000000000000000000000000000000000000000000000000000000007200000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000002a000000000000000000000000000000000000000000000000000000000000002400000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000014000000000000000000000000000000000000000000000000000000000000015490000000000000000000000000000000000000000000000000000000001745f2e0000000000000000000000000000000000000000000000000000000000120f460000000000000000000000000000000000000000000000000000000024f4730000000000000000000000000000000000000000000000000000000000007b80780000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000007d000000000000000000000000000000000000000000000000000000000000007d00000000000000000000000005707b946ee64bd750a587261ce36ec7024f3088b00000000000000000000000000000000000000000000000000000000007b80780000000000000000000000000000000000000000000000000000000000091da1000000000000000000000000000000000000000000000000000000006a7c1de9000000000000000000000000d1a45eba17b05cc62b11e2b62b8a00651a79014c00000000000000000000000000000000000000000000000000000000000000c00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000004100656d023165379b3e173bec991146f091f2b12022b28a125cdef7e3309effa2fea35c915f8c893958b52cbcfd0ab17c7b99865e82a7b4706d3caad398e8a604020000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000002a00000000000000000000000000000000000000000000000000000000000000240000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000001400000000000000000000000000000000000000000000000000000000000120f460000000000000000000000000000000000000000000000000000000024f47300000000000000000000000000000000000000000000000000000000000000154900000000000000000000000000000000000000000000000000000000017011f4000000000000000000000000eb5c5a71bff6ba9bff2c887155ed7283ff79696f0000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000007d000000000000000000000000000000000000000000000000000000000000007d00000000000000000000000005707b946ee64bd750a587261ce36ec7024f3088b000000000000000000000000eb5c5a71bff6ba9bff2c887155ed7283ff79696f000000000000000000000000000000000000000000000000000000000009305a000000000000000000000000000000000000000000000000000000006a7c1de9000000000000000000000000d1a45eba17b05cc62b11e2b62b8a00651a79014c00000000000000000000000000000000000000000000000000000000000000c000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000048010a450a003241f50213c411c0f42d9a7fd93d4ebb99d49f153476bf83416af40adf5c021458bd049f92b8ecb0dd030d118a3365b9cd9923ddfcf01204dea8e921418748ed614a1b00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000024f4730000000000000000000000000000000000000000000000000000000000",
    "estimate": true,
    "block": "39156482"
}
```
The block in the request body is `(the block when the transaction actually occurred) - 1` in order to represent the state before the transaction made its modifications. Of course, it is possible that in the same block and before the transaction consensus timestamp some other transactions might have influenced the state and altered the result. However, there are many examples like this. They can be found here: https://github.com/hiero-ledger/hiero-mirror-node/actions/runs/31577947233/job/94054230139?pr=13988

### Steps to reproduce

Run the "Gas Estimate Accuracy" workflow.

### Additional context

_No response_

### Hedera network

testnet

### Version

v0.161.0

### Operating system

None

### PR DESCRIPTION
**Pull request by IvanKavaldzhiev**
**Description**:
This PR improves transaction simulation accuracy by:

- Execute historical /contracts/call and opcode replay at the requested block’s consensus time instead of wall-clock Instant.now(), so transactionValidStart and TransactionExecutor.execute(...) match the original handle time. 
- Load system files at that same timestamp, so simulations use historical fees and exchange rates, not “now”.
- Bind BlockStreamInfo to the record file (blockNumber = index - 1, blockTime / lastHandleTime = consensus start) so exchange rates do not fall back to genesis config during historical calls.
- Enhnace the gas-accuracy-estimate script, so that if a simulation gas comparison fails with block: block_number - 1, there is a retry with block: block_number. It might be the case that the consensusTimestamp is near a full hour, making the CN execution running on the correct, up-to-date system files, while the mirror-node simulation runs using the system files from the previous hour (due to using a previous block number). This change also reduces part of the reported failures.

These changes should improve the accuracy for gas estimations and fix the Ethereum transactions incorrect gas estimates for most cases.

**Related issue(s)**:

Fixes #14064 

**Notes for reviewer**:
The gas-accuracy-estimate script was run against mainnet and this PR changes. Out of 15009 contract results, only 79 cases showed failure. Some of the failures are expected as there are transactions inside the block, which a given transaction depends on and we are missing this exact state in the simulations. The other cases can be further explored and fixed in smaller targeted PRs.

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java [hunk 1]
```diff
@@ -34,6 +34,18 @@ public String getServiceName() {
...
+        }
+        final var recordFile = ContractCallContext.get().getRecordFile();
+        if (recordFile == null) {
+            return FALLBACK;
+        }
+        final var blockTime = Utils.convertToTimestamp(recordFile.getConsensusStart());
+        return BlockStreamInfo.newBuilder()
+                .blockNumber(recordFile.getIndex() - 1)
+                .blockTime(blockTime)
+                .lastHandleTime(blockTime)
```
Comment (steven-sheehy) [inline web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java]: This should probably be `recordFile.getConsensusEnd()`

File: web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java [hunk 2]
```diff
@@ -34,6 +34,18 @@ public String getServiceName() {
...
+            return FALLBACK;
+        }
+        final var recordFile = ContractCallContext.get().getRecordFile();
+        if (recordFile == null) {
+            return FALLBACK;
+        }
+        final var blockTime = Utils.convertToTimestamp(recordFile.getConsensusStart());
+        return BlockStreamInfo.newBuilder()
+                .blockNumber(recordFile.getIndex() - 1)
+                .blockTime(blockTime)
```
Comment (steven-sheehy) [inline web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java:46]: We should also set `blockEndTime` to be consensusEnd

File: web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java [hunk 3]
```diff
@@ -34,6 +34,18 @@ public String getServiceName() {
...
+        if (!ContractCallContext.isInitialized()) {
+            return FALLBACK;
+        }
+        final var recordFile = ContractCallContext.get().getRecordFile();
+        if (recordFile == null) {
+            return FALLBACK;
+        }
+        final var blockTime = Utils.convertToTimestamp(recordFile.getConsensusStart());
+        return BlockStreamInfo.newBuilder()
+                .blockNumber(recordFile.getIndex() - 1)
```
Comment (steven-sheehy) [inline web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java]: Are you sure this should be decremented by one? The `BlockInfoSingleton` only did so because of some nuance in processing that code, but `BlockStreamInfo` processing could be different.

File: web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java [hunk 4]
```diff
@@ -34,6 +34,18 @@ public String getServiceName() {
 
     @Override
     public BlockStreamInfo get() {
-        return cachedBlockStreamInfo;
+        if (!ContractCallContext.isInitialized()) {
+            return FALLBACK;
+        }
```
Comment (steven-sheehy) [inline web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java]: We don't generally check if isInitialized() everywhere we use ContractCallContext because otherwise we'd have a bunch of unnecessary dead code.

File: web3/src/main/java/org/hiero/mirror/web3/state/keyvalue/FileReadableKVState.java [hunk 5]
```diff
@@ -51,15 +51,14 @@ protected File readFromDataSource(@NonNull FileID key) {
         final var timestamp = ContractCallContext.get().getTimestamp();
         final var fileEntityId = toEntityId(key);
         final var fileId = fileEntityId.getId();
-        final var currentTimestamp = getCurrentTimestamp();
+        final var effectiveTimestamp = timestamp.orElseGet(() -> getCurrentTimestamp());
```
Comment (steven-sheehy) [inline web3/src/main/java/org/hiero/mirror/web3/state/keyvalue/FileReadableKVState.java]:
```suggestion
        final var effectiveTimestamp = timestamp.orElseGet(Utils::getCurrentTimestamp);
```

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java [hunk 6]
```diff
@@ -34,6 +34,18 @@ public String getServiceName() {
...
+        final var recordFile = ContractCallContext.get().getRecordFile();
+        if (recordFile == null) {
+            return FALLBACK;
+        }
+        final var blockTime = Utils.convertToTimestamp(recordFile.getConsensusStart());
```
Comment (xin-hedera) [inline web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java:42]: **Possible bug:** `blockTime` and `lastHandleTime` are both derived from `recordFile.getConsensusStart()` here. Compare with the unchanged sibling `BlockInfoSingleton.get()`, which uses `getConsensusStart()` only for `firstConsTimeOfCurrentBlock` and `getConsensusEnd()` for the analogous "last handled transaction" field (`consTimeOfLastHandledTxn`). Meanwhile the transaction actually being simulated in the same call executes at `recordFile.getConsensusEnd()` (via `ContractCallContext.getTimestamp()` in `TransactionExecutionService`). Setting `lastHandleTime` to `consensusStart` instead of `consensusEnd` looks like it feeds an artificially-earlier "last handled" timestamp into the embedded hedera-app logic than the transaction it's about to execute — which could undercut this PR's own gas-estimation-accuracy goal. Should `lastHandleTime` use `getConsensusEnd()` instead (like `BlockInfoSingleton.consTimeOfLastHandledTxn`)?

File: web3/src/main/java/org/hiero/mirror/web3/service/TransactionExecutionService.java [hunk 7]
```diff
@@ -161,19 +162,23 @@ private EvmTransactionResult handleFailedResult(
         }
     }
 
-    private TransactionBody.Builder defaultTransactionBodyBuilder(final CallServiceParameters params) {
+    private TransactionBody.Builder defaultTransactionBodyBuilder(
+            final CallServiceParameters params, final Instant consensusNow) {
         return TransactionBody.newBuilder()
                 .transactionID(TransactionID.newBuilder()
-                        .transactionValidStart(new Timestamp(Instant.now().getEpochSecond(), 0))
+                        .transactionValidStart(new Timestamp(consensusNow.getEpochSecond(), consensusNow.getNano()))
```
Comment (xin-hedera) [inline web3/src/main/java/org/hiero/mirror/web3/service/TransactionExecutionService.java:170]: **Possible correctness issue:** for historical calls, `consensusNow` is now deterministic (`recordFile.getConsensusEnd()`, via `getConsensusTimeFromContext()`) rather than `Instant.now()`. Combined with `getSenderAccountID` returning the fixed `systemEntity.treasuryAccount()` whenever the sender is zero/default (the common `eth_call`/`estimateGas` case without an explicit `from`), two concurrent requests replaying the same historical block with a default sender will now build byte-identical `TransactionID`s (`accountID` + `transactionValidStart`). Previously `Instant.now()` gave practical per-request uniqueness even with the fixed payer. I didn't find a consumer in the standalone dispatch path that dedups on `TransactionID` today, but this removes a uniqueness guarantee that could matter if that changes, or for any logging/metrics/tracing correlation keyed on it.

File: web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java [hunk 8]
```diff
@@ -34,6 +34,18 @@ public String getServiceName() {
 
     @Override
     public BlockStreamInfo get() {
-        return cachedBlockStreamInfo;
+        if (!ContractCallContext.isInitialized()) {
```
Comment (xin-hedera) [inline web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java]: **Cleanup:** this `!ContractCallContext.isInitialized()` guard is now duplicated in `TransactionExecutionService.getConsensusTimeFromContext()` as well, while sibling singletons like `BlockInfoSingleton`/`RunningHashesSingleton` have no equivalent guard at all. Two independent ad hoc copies of "is context initialized, else fall back" suggests the invariant (context/blockSupplier always bound before any state singleton is read) belongs centrally in `ContractCallContext` rather than being re-implemented at each call site.

File: web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java [hunk 9]
```diff
@@ -34,6 +34,18 @@ public String getServiceName() {
...
+            return FALLBACK;
+        }
+        final var recordFile = ContractCallContext.get().getRecordFile();
+        if (recordFile == null) {
+            return FALLBACK;
+        }
+        final var blockTime = Utils.convertToTimestamp(recordFile.getConsensusStart());
+        return BlockStreamInfo.newBuilder()
```
Comment (xin-hedera) [inline web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java:44]: **Efficiency (minor):** `get()` rebuilds a new `BlockStreamInfo`/`Timestamp` on every invocation. This `SingletonState` getter can be called many times during a single EVM execution; previously this was a plain cached-field read. Worth considering computing this once per `ContractCallContext`/request and caching it there instead of reallocating on every call.

File: web3/src/main/java/org/hiero/mirror/web3/service/TransactionExecutionService.java [hunk 10]
```diff
@@ -200,8 +205,19 @@ private TransactionBody buildContractCallTransactionBody(
...
 
-    private TransactionBody buildEthereumTransactionBody(final ContractDebugParameters params) {
-        final var txnBody = defaultTransactionBodyBuilder(params)
+    private Instant getConsensusTimeFromContext() {
+        if (!ContractCallContext.isInitialized()) {
+            return Instant.now();
+        }
+        return ContractCallContext.get()
+                .getTimestamp()
+                .map(nanos -> Instant.ofEpochSecond(0L, nanos))
```
Comment (xin-hedera) [inline web3/src/main/java/org/hiero/mirror/web3/service/TransactionExecutionService.java]: **Reuse (minor):** this manual `Instant.ofEpochSecond(0L, nanos)` conversion duplicates the nanos-to-time idiom already expressed in `Utils.convertToTimestamp` / `SystemFileLoader`. Not a bug, but a third independent copy of this conversion increases the chance it drifts if the nanos-since-epoch convention ever needs to change.

File: web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java (hunk 3 above)
Comment (xin-hedera) [inline web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java]: **Altitude (minor):** `recordFile.getIndex() - 1` duplicates the same convention already in `BlockInfoSingleton.java` (`.lastBlockNumber(recordFile.getIndex() - 1)`), but that file has an explanatory comment ("Library internally increments last by one for current") and this one doesn't. Worth mirroring the comment here, or centralizing the offset, so the rationale isn't only documented in one of the two places that rely on it.

File: web3/src/test/java/org/hiero/mirror/web3/service/TransactionExecutionServiceTest.java [hunk 11]
```diff
@@ -165,6 +166,44 @@ void testExecuteContractCallSuccess(String senderAddressHex) {
         assertThat(result.functionResult().errorMessage()).isNull();
     }
 
+    @Test
+    void executeUsesHistoricalGetConsensusTimeFromContext() {
```
Comment (xin-hedera) [inline web3/src/test/java/org/hiero/mirror/web3/service/TransactionExecutionServiceTest.java:173]: **Test gap:** this new test only exercises the branch where `ContractCallContext.timestamp` is explicitly set (the opcode-replay-style path). The primary historical `/contracts/call` path — a `block` param with no explicit timestamp, which falls back to `recordFile.getConsensusEnd()` via `ContractCallContext.getTimestampOrDefaultFromRecordFile()` — isn't covered by a test asserting the consensus time passed to `executor.execute(...)`. That's the path most affected by the `consensusStart`/`consensusEnd` question raised on `BlockStreamInfoSingleton`.

**Comment by xin-hedera** [comment]
**Edge case (couldn't attach inline since `ContractDebugService.java` isn't part of this diff, only a caller of the changed code):** in `ContractDebugService.processOpcodeCall` ([ContractDebugService.java:51](https://github.com/hiero-ledger/hiero-mirror-node/blob/b46c6f1e5ecf5976e597be102eb281987639f060/web3/src/main/java/org/hiero/mirror/web3/service/ContractDebugService.java#L51)), `ctx.setTimestamp(Optional.of(params.getConsensusTimestamp() - 1))` now also drives `SystemFileLoader.load`'s `roundDownToHour(...)` for exchange-rate/fee-schedule lookups during opcode replay via `TransactionExecutionService.getConsensusTimeFromContext()` — previously system files always loaded at wall-clock "now", so this historical value was never used for them. If a replayed transaction's real consensus timestamp lands exactly on an hour boundary (nanos component a multiple of `NANOS_PER_HOUR`), subtracting 1ns rounds the effective timestamp into the *previous* hour, so the debug replay could pick up the wrong (previous) exchange-rate/fee-schedule file version versus what was actually in effect for that transaction. Narrow, but the code doesn't guard against it.

File: web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java (hunk 1 above)
Comment (IvanKavaldzhiev) [reply web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java]: Fixed

File: web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java (hunk 2 above)
Comment (IvanKavaldzhiev) [reply web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java:46]: Fixed

File: web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java (hunk 4 above)
Comment (IvanKavaldzhiev) [reply web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java]: Removed

File: web3/src/main/java/org/hiero/mirror/web3/state/keyvalue/FileReadableKVState.java (hunk 5 above)
Comment (IvanKavaldzhiev) [reply web3/src/main/java/org/hiero/mirror/web3/state/keyvalue/FileReadableKVState.java]: Fixed

File: web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java (hunk 6 above)
Comment (IvanKavaldzhiev) [reply web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java:42]: Fixed

File: web3/src/main/java/org/hiero/mirror/web3/service/TransactionExecutionService.java (hunk 7 above)
Comment (IvanKavaldzhiev) [reply web3/src/main/java/org/hiero/mirror/web3/service/TransactionExecutionService.java]: I restored back the usage of Instant.now().

File: web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java (hunk 8 above)
Comment (IvanKavaldzhiev) [reply web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java]: Removed

File: web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java (hunk 9 above)
Comment (IvanKavaldzhiev) [reply web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java:44]: Fixed

File: web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java (hunk 3 above)
Comment (IvanKavaldzhiev) [reply web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java]: Yes, seems the decrement is not needed for this class.

Comment (IvanKavaldzhiev) [reply web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java]: Removed

File: web3/src/test/java/org/hiero/mirror/web3/service/TransactionExecutionServiceTest.java (hunk 11 above)
Comment (IvanKavaldzhiev) [reply web3/src/test/java/org/hiero/mirror/web3/service/TransactionExecutionServiceTest.java:173]: Added more test cases

**Comment by IvanKavaldzhiev** [comment]
> **Edge case (couldn't attach inline since `ContractDebugService.java` isn't part of this diff, only a caller of the changed code):** in `ContractDebugService.processOpcodeCall` ([ContractDebugService.java:51](https://github.com/hiero-ledger/hiero-mirror-node/blob/b46c6f1e5ecf5976e597be102eb281987639f060/web3/src/main/java/org/hiero/mirror/web3/service/ContractDebugService.java#L51)), `ctx.setTimestamp(Optional.of(params.getConsensusTimestamp() - 1))` now also drives `SystemFileLoader.load`'s `roundDownToHour(...)` for exchange-rate/fee-schedule lookups during opcode replay via `TransactionExecutionService.getConsensusTimeFromContext()` — previously system files always loaded at wall-clock "now", so this historical value was never used for them. If a replayed transaction's real consensus timestamp lands exactly on an hour boundary (nanos component a multiple of `NANOS_PER_HOUR`), subtracting 1ns rounds the effective timestamp into the _previous_ hour, so the debug replay could pick up the wrong (previous) exchange-rate/fee-schedule file version versus what was actually in effect for that transaction. Narrow, but the code doesn't guard against it.

Fixed by keeping an additional timestamp in OpcodeContext. Now 2 timestamps are saved - 1 for the previousBlock, 1 for the actual consensusTimestamp. In this way based on the data type that is loaded the correct timestamp is used.

File: web3/src/main/java/org/hiero/mirror/web3/service/TransactionExecutionService.java (hunk 10 above)
Comment (IvanKavaldzhiev) [reply web3/src/main/java/org/hiero/mirror/web3/service/TransactionExecutionService.java]: I don't think this comment is relevant anymore.

File: web3/src/main/java/org/hiero/mirror/web3/repository/EntityRepository.java [hunk 12]
```diff
@@ -161,4 +163,18 @@ order by lower(timestamp_range) desc
...
+    @Query(value = """
+                    select max(id)
+                    from (
+                        select id
+                        from entity
+                        where created_timestamp <= ?1
+                        union all
+                        select id
+                        from entity_history
+                        where created_timestamp <= ?1
```
Comment (steven-sheehy) [inline web3/src/main/java/org/hiero/mirror/web3/repository/EntityRepository.java]: This query could be slow because created_timestamp is not an index, nor can we add it as one.

File: web3/src/main/java/org/hiero/mirror/web3/service/ContractDebugService.java [hunk 13]
```diff
@@ -48,6 +48,7 @@ public ContractDebugService(
     public OpcodesProcessingResult processOpcodeCall(
             final @Valid ContractDebugParameters params, final OpcodeContext opcodeContext) {
         ContractCallContext ctx = ContractCallContext.get();
+        opcodeContext.setConsensusTimestamp(Optional.of(params.getConsensusTimestamp()));
         ctx.setTimestamp(Optional.of(params.getConsensusTimestamp() - 1));
```
Comment (steven-sheehy) [inline web3/src/main/java/org/hiero/mirror/web3/service/ContractDebugService.java]: Do we really need a separate timestamp when it's the same plus one?

File: web3/src/main/java/org/hiero/mirror/web3/service/TransactionExecutionService.java [hunk 14]
```diff
@@ -161,19 +167,35 @@ private EvmTransactionResult handleFailedResult(
...
+                                new Timestamp(transactionValidStart.getEpochSecond(), transactionValidStart.getNano()))
                         .accountID(getSenderAccountID(params))
                         .build())
                 .nodeAccountID(EntityIdUtils.toAccountId(systemEntity.treasuryAccount()))
                 .transactionValidDuration(TRANSACTION_DURATION);
     }
 
+    /**
+     * Subtracts a process-wide sequence from {@code consensusNow} so concurrent historical calls with the same
+     * default payer do not share a TransactionID, while remaining inside the 15s validity window.
```
Comment (steven-sheehy) [inline web3/src/main/java/org/hiero/mirror/web3/service/TransactionExecutionService.java]: What happens when they share the same transaction ID? Does the CN library reject as duplicate or can we control that behavior?

File: web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java [hunk 15]
```diff
@@ -34,6 +34,25 @@ public String getServiceName() {
...
+        }
+        final var blockTime = Utils.convertToTimestamp(recordFile.getConsensusStart());
+        final var blockEndTime = Utils.convertToTimestamp(recordFile.getConsensusEnd());
+        final var blockStreamInfo = BlockStreamInfo.newBuilder()
+                .blockNumber(recordFile.getIndex())
+                .blockTime(blockTime)
+                .lastHandleTime(blockEndTime)
+                .blockEndTime(blockEndTime)
+                .build();
+        context.setBlockStreamInfo(blockStreamInfo);
```
Comment (steven-sheehy) [inline web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java]: Is this actually invoked multiple times per contract call? Would be easy to debug and verify. Seems like CN would invoke it once at most and the cache is pointless.

File: web3/src/main/java/org/hiero/mirror/web3/repository/EntityRepository.java (hunk 12 above)
Comment (IvanKavaldzhiev) [reply web3/src/main/java/org/hiero/mirror/web3/repository/EntityRepository.java]: I removed the query. The use case of it shouldn't impact the historical behaviour of transactions. We just need a unique not existing next numeric value.

File: web3/src/main/java/org/hiero/mirror/web3/service/ContractDebugService.java (hunk 13 above)
Comment (IvanKavaldzhiev) [reply web3/src/main/java/org/hiero/mirror/web3/service/ContractDebugService.java]: I simplified the code and deleted the field.

File: web3/src/main/java/org/hiero/mirror/web3/service/TransactionExecutionService.java (hunk 14 above)
Comment (IvanKavaldzhiev) [reply web3/src/main/java/org/hiero/mirror/web3/service/TransactionExecutionService.java]: The hedera-app shouldn't block such transactions as each transaction as we set a default RecordCacheService without any entries. I simplified the code, without worrying about duplicate TransactionIDs for now.

File: web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java (hunk 15 above)
Comment (IvanKavaldzhiev) [reply web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java]: The class is invoked multiple times but debugging showed that a lot of times the FALLBACK is returned. The class might be invoked early in the lifecycle and even though we set a RecordFile it's not always present during this method execution, so we can skip the cache. Removed it.

File: web3/src/main/java/org/hiero/mirror/web3/service/TransactionExecutionService.java [hunk 16]
```diff
@@ -58,6 +60,9 @@
 public class TransactionExecutionService {
 
     private static final Duration TRANSACTION_DURATION = new Duration(15);
+    private static final long MAX_TRANSACTION_VALID_START_OFFSET_NANOS =
+            TRANSACTION_DURATION.seconds() * 1_000_000_000L - 1;
+    private static final AtomicLong TRANSACTION_ID_SEQUENCE = new AtomicLong();
```
Comment (steven-sheehy) [inline web3/src/main/java/org/hiero/mirror/web3/service/TransactionExecutionService.java]: These are now unused

File: web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java [hunk 17]
```diff
@@ -34,6 +34,19 @@ public String getServiceName() {
 
     @Override
     public BlockStreamInfo get() {
-        return cachedBlockStreamInfo;
+        final var context = ContractCallContext.get();
+        final var recordFile = context.getRecordFile();
+        if (recordFile == null) {
+            // Do not cache the fallback: the record file may be bound later on the same context (warmup, tests).
```
Comment (steven-sheehy) [inline web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java]: Caching comment is now irrelevant.

File: web3/src/main/java/org/hiero/mirror/web3/state/singleton/EntityIdSingleton.java [hunk 18]
```diff
@@ -43,6 +43,7 @@ public EntityNumber get() {
                 .getConfigData(HederaConfig.class)
                 .firstUserEntity();
 
+        final var timestamp = context.getTimestamp();
```
Comment (steven-sheehy) [inline web3/src/main/java/org/hiero/mirror/web3/state/singleton/EntityIdSingleton.java]: unused?

File: web3/src/test/java/org/hiero/mirror/web3/state/singleton/EntityIdSingletonTest.java [hunk 19]
```diff
@@ -25,6 +26,9 @@ class EntityIdSingletonTest {
     @Mock
     private EntityRepository entityRepository;
 
+    @Mock
+    private CallServiceParameters callServiceParameters;
```
Comment (steven-sheehy) [inline web3/src/test/java/org/hiero/mirror/web3/state/singleton/EntityIdSingletonTest.java]: Unused?

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: web3/src/main/java/org/hiero/mirror/web3/service/TransactionExecutionService.java (hunk 16 above)
Comment (IvanKavaldzhiev) [reply web3/src/main/java/org/hiero/mirror/web3/service/TransactionExecutionService.java]: Removed

File: web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java (hunk 17 above)
Comment (IvanKavaldzhiev) [reply web3/src/main/java/org/hiero/mirror/web3/state/singleton/BlockStreamInfoSingleton.java]: Removed

File: web3/src/main/java/org/hiero/mirror/web3/state/singleton/EntityIdSingleton.java (hunk 18 above)
Comment (IvanKavaldzhiev) [reply web3/src/main/java/org/hiero/mirror/web3/state/singleton/EntityIdSingleton.java]: Removed

File: web3/src/test/java/org/hiero/mirror/web3/state/singleton/EntityIdSingletonTest.java (hunk 19 above)
Comment (IvanKavaldzhiev) [reply web3/src/test/java/org/hiero/mirror/web3/state/singleton/EntityIdSingletonTest.java]: Removed

**Review by steven-sheehy — APPROVED** [review]
